src/libraries/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePartPiece.cs

```diff
@@ -0,0 +1,214 @@
+            // write >10 part pieces. These should be parsed
+            success = success
+                && searchPosition > 1
+                && char.IsDigit(path[searchPosition - 1]);
+
+            if (success)
+            {
+                int digitStart;
+
+                // Iterate backwards, character by character, until we find a non-digit character.
+                for (digitStart = searchPosition; digitStart > 1 && char.IsDigit(path[digitStart - 1]); digitStart--)
+                    ;
+
+                success = int.TryParse(path.Substring(digitStart, searchPosition - digitStart), out pieceNumber);
+                if (success)
+                {
+                    searchPosition = digitStart;
+                }
+            }
+
+            // Stage 4: locate and remove the separator directly after the piece name
+            success = success
+                && searchPosition > 1
+                && path[searchPosition - 1] == '['
+                && path[searchPosition - 2] == '/';
+
+            if (success)
+            {
+                searchPosition -= 2;
+
+                // Stage 5: extract the piece name and validate it
+                if (searchPosition > 0)
+                {
+                    int searchOffset = path[0] == '/' ? 1 : 0;
+
+                    prefixName = path.Substring(searchOffset, searchPosition - searchOffset);
+
+                    success = success
+                        && Uri.TryCreate(ZipPackage.GetOpcNameFromZipItemName(prefixName), UriKind.Relative, out Uri? unvalidatedPartUri)
+                        && PackUriHelper.TryValidatePartUri(unvalidatedPartUri, out partUri);
+                }
+                else
+                {
+                    success = false;
+                }
+            }
+
+            return success;
+        }
+
+        internal static ZipPackagePartPiece Create(ZipArchive zipArchive, PackUriHelper.ValidatedPartUri? partUri, string prefixName, int pieceNumber, bool isLastPiece)
+        {
+            string newPieceFileName = FormattableString.Invariant($"{prefixName}/[{pieceNumber:D}]{(isLastPiece ? ".last" : string.Empty)}.piece");
+            ZipArchiveEntry newPieceEntry = zipArchive.CreateEntry(newPieceFileName);
+
+            return new ZipPackagePartPiece(newPieceEntry, partUri, prefixName, pieceNumber, isLastPiece);
+        }
+        #endregion
+
+        #region Internal Constructors
+        internal ZipPackagePartPiece(ZipArchiveEntry zipArchiveEntry, PackUriHelper.ValidatedPartUri? partUri, string prefixName, int pieceNumber, bool isLastPiece)
+        {
+            if (zipArchiveEntry == null)
+                throw new ArgumentNullException(nameof(zipArchiveEntry));
+            if (prefixName == null)
+                throw new ArgumentNullException(nameof(prefixName));
+            if (pieceNumber < 0)
+                throw new ArgumentOutOfRangeException(nameof(pieceNumber));
```

Comment:
Do we care if `prefixName` is an empty string? In case we need its exception to be `ArgumentNullException.ThrowIfNull` instead.

```suggestion
            ArgumentNullException.ThrowIfNull(zipArchiveEntry, nameof(zipArchiveEntry));
            ArgumentException.ThrowIfNullOrEmpty(prefixName, nameof(prefixName));
            ArgumentOutOfRangeException.ThrowIfNegative(pieceNumber, nameof(pieceNumber));
```

Reply:
`prefixName` can't be empty - the prefix is the parent directory, and part pieces in the root directory are invalid. The `InvalidPartPiecesAreNotParsable` test validates that these are blocked.

I've switched everything to use the exception helper methods (although System.IO.Packaging is built for .NET Standard, so I've left the originals in situ for that.)